File: main/src/models/lecturer.model.ts
import { LoginTypeEnum } from '@/enums/login-type.enum';
import { BaseModel } from '@/models/base.model';
import { Course } from '@/models/course.model';
import { Enrollment } from '@/models/enrollment.model';
import { Role } from '@/models/role.model';
import { Entity, PrimaryGeneratedColumn, Column, OneToMany, JoinColumn, ManyToOne, Unique } from 'typeorm';

@Unique(['phoneNumber'])
@Unique(['email'])
@Entity('lecturers')
export class Lecturer extends BaseModel {
  @PrimaryGeneratedColumn('uuid')
  id!: string;

  @Column({ length: 50 })
  name!: string;

  @Column({ length: 70, nullable: true })
  email!: string;

  @Column({ nullable: false, default: false, name: 'email_verified' })
  emailVerified!: boolean;

  @Column({ length: 15, nullable: true, name: 'phone_number' })
  phoneNumber!: string;

  @Column({ length: 100 })
  address!: string;

  @Column({ type: 'text', nullable: true })
  bio!: string;

  // @Column({ type: 'enum', enum: LoginTypeEnum, default: LoginTypeEnum.email, name: 'login_type' })
  // loginType!: LoginTypeEnum;

  @Column({ length: 150 })
  password!: string;

  @Column({ nullable: false, default: false, name: 'is_approved' })
  isApproved!: boolean;

  @Column({ type: 'text', nullable: true })
  avatar?: string;

  @Column({ length: 100, nullable: true, name: 'short_description' })
  shortDescription?: string;

  @Column({ type: 'text', nullable: true, name: 'example_video' })
  exampleVideo?: string;

  @Column({ name: 'social_link', type: 'text', nullable: true })
  socialLink?: string;

  @Column({ name: 'teaching_topic', type: 'varchar', length: 50, nullable: true })
  teachingTopic?: string;

  @Column({ name: 'teaching_experience', type: 'text', nullable: true })
  teachingExperience?: string;

  @OneToMany(() => Course, (course) => course.lecturer)
  courses!: Promise<Course[]>;

  @Column({ nullable: true, name: 'role_id' })
  roleId!: string;

  @ManyToOne(() => Role, (role) => role.employees)
  @JoinColumn({ name: 'role_id' })
  role!: Promise<Role>;
}

File: lib/models/lecturer_model.dart
class Lecturer {
  final String? status;
  final int? code;
  final bool? success;
  final String? message;
  final LecturerData? data;
  final dynamic errors;

  Lecturer({
    this.status,
    this.code,
    this.success,
    this.message,
    this.data,
    this.errors,
  });

  Lecturer.fromJson(Map<String, dynamic> json)
      : status = json['status'] as String?,
        code = json['code'] as int?,
        success = json['success'] as bool?,
        message = json['message'] as String?,
        data = (json['data'] as Map<String,dynamic>?) != null ? LecturerData.fromJson(json['data'] as Map<String,dynamic>) : null,
        errors = json['errors'];

  Map<String, dynamic> toJson() => {
    'status' : status,
    'code' : code,
    'success' : success,
    'message' : message,
    'data' : data?.toJson(),
    'errors' : errors
  };
}

class LecturerData {
  final String? id;
  final String? name;
  final String? email;
  final String? phoneNumber;
  final String? address;
  final String? bio;
  final bool? isApproved;

  LecturerData({
    this.id,
    this.name,
    this.email,
    this.phoneNumber,
    this.address,
    this.bio,
    this.isApproved,
  });

  LecturerData.fromJson(Map<String, dynamic> json)
      : id = json['id'] as String?,
        name = json['name'] as String?,
        email = json['email'] as String?,
        phoneNumber = json['phoneNumber'] as String?,
        address = json['address'] as String?,
        bio = json['bio'] as String?,
        isApproved = json['isApproved'] as bool?;

  Map<String, dynamic> toJson() => {
    'id' : id,
    'name' : name,
    'email' : email,
    'phoneNumber' : phoneNumber,
    'address' : address,
    'bio' : bio,
    'isApproved' : isApproved
  };
}